src/Microsoft.SqlTools.ServiceLayer/Profiler/ProfilerService.cs

```diff
@@ -1,420 +1,464 @@
+                }
+                catch { }
+
+                // create a new XEvent session and Profiler session, if it doesn't exist
+                xeSession ??= this.XEventSessionFactory.CreateXEventSession(parameters.Template.CreateStatement, parameters.SessionName, connInfo);
+
+                // start monitoring the profiler session
+                monitor.StartMonitoringSession(parameters.OwnerUri, xeSession);
+
+                var result = new CreateXEventSessionResult();
+                await requestContext.SendResult(result);
+
+                SessionCreatedNotification(parameters.OwnerUri, parameters.SessionName, parameters.Template.Name);
+            }
+        }
+
+        /// <summary>
+        /// Handle request to start a profiling session
+        /// </summary>
+        internal async Task HandleStartProfilingRequest(StartProfilingParams parameters, RequestContext<StartProfilingResult> requestContext)
+        {
+            if (parameters.SessionType == ProfilingSessionType.LocalFile)
+            {
+                await StartLocalFileSession(parameters, requestContext);
+                return;
+            }
+            ConnectionInfo connInfo;
+            ConnectionServiceInstance.TryFindConnection(
+                parameters.OwnerUri,
+                out connInfo);
+            if (connInfo != null)
+            {
+                // create a new XEvent session and Profiler session
+                var xeSession = this.XEventSessionFactory.GetXEventSession(parameters.SessionName, connInfo);
+
+                // start monitoring the profiler session
+                monitor.StartMonitoringSession(parameters.OwnerUri, xeSession);
+
+                var result = new StartProfilingResult() { CanPause = true, UniqueSessionId = xeSession.Id.ToString() };
+                await requestContext.SendResult(result);
+            }
+            else
+            {
+                throw new ProfilerException(SR.ProfilerConnectionNotFound);
+            }
+        }
+
+        private async Task StartLocalFileSession(StartProfilingParams parameters, RequestContext<StartProfilingResult> requestContext)
+        {
+            var xeSession = XEventSessionFactory.OpenLocalFileSession(parameters.SessionName);
+            monitor.StartMonitoringSession(parameters.OwnerUri, xeSession);
+            xeSession.Start();
+            var result = new StartProfilingResult() { UniqueSessionId = xeSession.Id.ToString(), CanPause = false };
+            await requestContext.SendResult(result);
+        }
+
+        public IXEventSession OpenLocalFileSession(string filePath)
+        {
+            return new ObservableXEventSession(() =>
+            {
+                Task<IXEventFetcher> task1 = initIXEventFetcher(filePath);
+
+                var iXEventFetcher = task1.GetAwaiter().GetResult();
+                return iXEventFetcher;
+            }, new SessionId(filePath));
+        }

```

Comment:
this feels like too much `Task` management at this level.
Can we not just pass `() => new XEFileEventStreamer(filePath)` as the func, and have `XeStreamObservable.Start` handle the exception from the func, and then use `Task.FromException.ContinueWith` instead of calling `ReadEventStream.ContinueWith`?

Reply:
Updated it appropriately.